code/modules/reagents/chemistry/items.dm

```diff
@@ -162,3 +131,198 @@
+			container.reagents.expose_temperature(get_temperature())
+			to_chat(user, "<span class='notice'>You heat up the [src].</span>")
+			playsound(user.loc, 'sound/chemistry/heatdam.ogg', 50, TRUE)
+			return
+	else if(isitem(target))
+		var/obj/item/item = target
+		if(item.heat > 1000)
+			set_lit(TRUE)
+			user.visible_message("<span class='notice'>[user] lights up the [src].</span>")
+
+/obj/item/burner/update_icon_state()
+	icon_state = "[initial(icon_state)][lit ? "-on" : ""]"
+
+
+/obj/item/burner/proc/set_lit(new_lit)
+	if(lit == new_lit)
+		return
+	lit = new_lit
+	if(lit)
+		force = 5
+		damtype = BURN
+		hitsound = 'sound/items/welder.ogg'
+		attack_verb_continuous = string_list(list("burns", "sings"))
+		attack_verb_simple = string_list(list("burn", "sing"))
+		START_PROCESSING(SSobj, src)
+	else
+		hitsound = "swing_hit"
+		force = 0
+		attack_verb_continuous = null //human_defense.dm takes care of it
+		attack_verb_simple = null
+		STOP_PROCESSING(SSobj, src)
+	set_light_on(lit)
+	update_icon()
+
+/obj/item/burner/extinguish()
+	set_lit(FALSE)
+
+/obj/item/burner/attack_self(mob/living/user)
+	. = ..()
+	if(.)
+		return
+	if(lit)
+		set_lit(FALSE)
+		user.visible_message("<span class='notice'>[user] snuffs out [src]'s flame.</span>")
+
+/obj/item/burner/attack(mob/living/carbon/M, mob/living/carbon/user)
+	if(lit && M.IgniteMob())
+		message_admins("[ADMIN_LOOKUPFLW(user)] set [key_name_admin(M)] on fire with [src] at [AREACOORD(user)]")
+		log_game("[key_name(user)] set [key_name(M)] on fire with [src] at [AREACOORD(user)]")
+	return ..()
+
+/obj/item/burner/process()
+	var/current_heat = 0
+	var/number_of_burning_reagents = 0
+	for(var/datum/reagent/reagent as anything in reagents.reagent_list)
+		reagent.burn(reagents) //burn can set temperatures of reagents
+		if(!isnull(reagent.burning_temperature))
+			current_heat += reagent.burning_temperature
+			number_of_burning_reagents += 1
+			reagents.remove_reagent(reagent.type, reagent.burning_volume)
+			continue
+
+	if(!number_of_burning_reagents)
+		set_lit(FALSE)
+		heat = 0
+		return
+	open_flame()
+	current_heat /= number_of_burning_reagents
+	heat = current_heat
+
+/obj/item/burner/get_temperature()
+	return lit * heat
+
+/obj/item/burner/oil
+	name = "Oil burner"
+	reagent_type = /datum/reagent/fuel/oil
+	grind_results = list(/datum/reagent/fuel/oil = 5, /datum/reagent/silicon = 10)
+
+/obj/item/burner/fuel
+	name = "Fuel burner"
+	reagent_type = /datum/reagent/fuel
+	grind_results = list(/datum/reagent/fuel = 5, /datum/reagent/silicon = 10)
+
+/obj/item/thermometer
+	name = "thermometer"
+	desc = "A thermometer for checking a beaker's temperature"
+	icon_state = "thermometer"
+	icon = 'icons/obj/chemical.dmi'
+	item_flags = NOBLUDGEON
+	w_class = WEIGHT_CLASS_TINY
+	grind_results = list(/datum/reagent/mercury = 5)
+	///The beaker that this object is attached to, so we know where we are when it's added to something.
+	var/obj/item/reagent_containers/attached_beaker
```

Comment:
QDEL_NULL this variable on the thermometer's destruction.

This is a general principle: always clean your references while in `Destroy()` if you cannot guarantee they'll be `null` by then.